R/future_promise.R

```diff
@@ -281,7 +300,194 @@ future_promise_queue <- local({
+#'
+#' Typically, the expression to be executed can be delayed in a [promise()] before submission to a \pkg{future} worker.
+#'
+#' @section `promise` vs `future`:
+#'
+#' With future blocking on the main R session, this prevents values from being changed.  It is known that evironments can change their values unexpectedly. As with [promise()], it is recommended to pull out important variables and force them to a local value before creating your `future_promise()`.
+#'
+#' For example, the environtment `env` below will have its value `a` changed before it is used.
+#' ```r
+#' {
+#'   env <- new.env()
+#'   env$a <- 1
+#'
+#'   promise_resolve(TRUE) %...>%
+#'     { Sys.sleep(1); env$a } %...>%
+#'     { print(.) }
+#'   env$a <- 2
+#'   print("done")
+#' }
+#' #> [1] "done"
+#' #> [1] 2
+#' ```
+#'
+#' To address this, we can capture the value `a` or turn the environment into a `list()`.
+#' ```r
+#' {
+#'   env <- new.env()
+#'   env$a <- 1
+#'
+#'   a_val <- env$a
+#'   promise_resolve(TRUE) %...>%
+#'     { Sys.sleep(1); a_val } %...>%
+#'     { print(.) }
+#'   env$a <- 2
+#'   print("done")
+#' }
+#' #> [1] "done"
+#' #> [1] 1
+#' ```
+#'
+#' When using `future_promise()`, be sure to scope your variables if the context can change. `future_promise()` will behave more like a [promise()] that executes using [future::future()].
+#'
+#' Changing context example:
+#' ```r
+# items <- list()
+# for (i in 1:10) {
+#   items <- c(items, list(
+#     promise(function(resolve, reject) {
+#       later::later(function() { resolve(i) })
+#     })
+#   ))
+# }
+# promise_all(.list = items) %...>%
+#   { print(unlist(.)) }
+# #> [1] 10 10 10 10 10 10 10 10 10 10
+#' ```
+#'
+#' Scoped context example:
+#' ```r
+#' lapply(1:10, function(i) {
+#'   promise(function(resolve, reject) {
+#'     later::later(function() { resolve(i) })
+#'   })
+#' }) %>%
+#'   promise_all(.list = .) %...>%
+#'   { print(unlist(.)) }
+#' #> [1]  1  2  3  4  5  6  7  8  9 10
+#' ```
+#'
+#' @describeIn future_promise Creates a [promise()] that will execute the `expr` using [future::future()].
+#' @inheritParams future::future
+#' @param ... extra parameters provided to future
+#' @param queue A queue that is used to schedule work to be done using [future::future()].  See the details for more information.
+#' @return a [promise()] object that will return the result of the calculated `expr`.
+#' @examples
+#' \donttest{local({
+#'   ## Setup ##
+#'
+#'   # Reset future plan on exit
+#'   old_plan <- future::plan()
+#'   on.exit({future::plan(old_plan)}, add = TRUE)
+#'
+#'   # Set up a plan with 2 future workers
+#'   future::plan(future::multisession(workers = 2))
```

Comment:
Wrap in a `with_plan()` so you can run parts of the code below and get the output in a readable fashion